### PR TITLE
[scaffolding-chef-infra] WIP: refactor scaffolding to template files

### PR DESCRIPTION
**Pull request by echohack**
Signed-off-by: David Echo <echohack@users.noreply.github.com>

This is a refactoring of the main `scaffolding` library in order to make more sense of how files get templated out at buildtime vs runtime.

This significantly reduces the error-prone method of making code changes inside of `EOF` statements, and also opens up future refactoring to other languages, faster patterns, and so on. It also makes it easier for us to apply linting tools for specific language problems (eg: ruby config files, json, bash, and so on)

In the process of this refactor, I was able to discover many subtle and edge-case kind of bugs that were undocumented, which demonstrates the large quality improvement that this provides.

Finally, this pattern makes writing CI tests much easier, and a large part of this refactor was done as a result of the discoveries made in the previous CI test work.

Minor note: the syntax of the attributes.json file is always red in linters and other tools because of the handlebars syntax. I'm not sure how we should properly handle that.

WIP: Please review this code for Linux. Based on that review, I will also be doing this work for Windows, which will be included in this same pull request.

The work for [scaffolding-chef-inspec] will be done separately, but this work should forge the path ahead, so that doing work in Powershell or for [scaffolding-chef-inspec] will be significantly easier, because I've already discovered many of the gotchas.

### CODE REVIEW

File: scaffolding-chef-infra/lib/linux/attributes.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{{#if cfg.attributes ~}}
```
Comment (mikemol) [inline scaffolding-chef-infra/lib/linux/attributes.json]: Am I reading this correctly in that the attributes are updatable via the habitat ring?

Comment (TheLunaticScripter) [reply scaffolding-chef-infra/lib/linux/attributes.json]: Yes, they are.

**Comment by echohack** [comment]
I'm going to go ahead and merge this, as it's been stale for a while and is necessary to move many of the other PRs sitting out there. This change does not immediately warrant a new release but the changes that follow it will push a new release out fairly soon.

File: scaffolding-chef-infra/lib/linux/client-chunk.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+cfg_env_path_prefix = {{cfg.env_path_prefix}}
```
[the remark this answers is not in the document]
Comment (echohack) [reply scaffolding-chef-infra/lib/linux/client-chunk.rb]: Hi @afiune 

I think you're right that this would be stronger with single quotes, as ruby will treat this as a string literal. Thanks for this feedback.

File: scaffolding-chef-infra/lib/linux/run.bash [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#!/bin/sh
+
+export SSL_CERT_FILE="{{ pkgPathFor "${scaffold_cacerts}" }}/ssl/cert.pem"
+export SSL_CERT_DIR="{{ pkgPathFor "${scaffold_cacerts}" }}/ssl/certs"
```
[the remark this answers is not in the document]
Comment (echohack) [reply scaffolding-chef-infra/lib/linux/run.bash]: This variable is templated at build time. We are using $scaffold_cacerts here as a template. Please see this code: https://github.com/chef/effortless/blob/master/scaffolding-chef-infra/lib/linux/scaffolding.sh#L52

The default is always set in the build. See here: https://github.com/chef/effortless/blob/master/scaffolding-chef-infra/lib/linux/scaffolding.sh#L17

File: scaffolding-chef-infra/lib/linux/client-chunk.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (afiune) [reply scaffolding-chef-infra/lib/linux/client-chunk.rb]: Stronger? I mean this is ruby, and ruby, as far as I know, ruby can’t understand strings without quotes.

Simple test with quotes:
```
irb(main):002:0> cfg_env_path_prefix ='abc'
=> "abc"
```
Simple test **WITHOUT** quotes:
```
irb(main):001:0> cfg_env_path_prefix = abc
Traceback (most recent call last):
        2: from /usr/local/bin/irb:11:in `<main>'
        1: from (irb):1
NameError (undefined local variable or method `abc' for main:Object)
```

File: scaffolding-chef-infra/lib/linux/run.bash (hunk 3 above)
[the remark this answers is not in the document]
Comment (afiune) [reply scaffolding-chef-infra/lib/linux/run.bash]: Nope. I'm sorry but that bash code you sent me is only checking if the variable exists, add it as a package dep `pkg_deps`, else you add "the default `core/cacerts`". 
```bash
  if [ -n "${scaffold_cacerts}" ]; then
    pkg_deps+=("${scaffold_cacerts}")
  else
    pkg_deps+=("core/cacerts")
  fi
```
**But you never set the variable to any default.**

File: scaffolding-chef-infra/lib/linux/client-chunk.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (echohack) [reply scaffolding-chef-infra/lib/linux/client-chunk.rb]: Thank you! @afiune I've got a PR open to fix the bug :) #141

File: scaffolding-chef-infra/lib/linux/run.bash (hunk 3 above)
[the remark this answers is not in the document]
Comment (echohack) [reply scaffolding-chef-infra/lib/linux/run.bash]: Ah I see, `export SSL_CERT_FILE="{{ pkgPathFor "${CFG_CACERTS:-core/cacerts}" }}/ssl/cert.pem"`

This is how it was setting the variable default previously. I understand how this was confusing because of the double-build that was using the heredoc as a method of compiling these variables together.

We'll need to set a default in the sed command so it can template out successfully.
